File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
answer = 0 
def DFS(startindex,sumcnt,n,numbers,target) :
    #종단 조건
    global answer
    
    if startindex == n-1 :
        #마지막이고
        if sumcnt == target :
            answer +=1 
            return
            
    
    if 0<=startindex+1<n :
        DFS(startindex+1,sumcnt+numbers[startindex+1],n,numbers,target)
        DFS(startindex+1,sumcnt-numbers[startindex+1],n,numbers,target)

        

def solution(numbers, target):
    n = len(numbers) 
    DFS(-1,0,n,numbers,target)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    public int answer = 0 ;
    public int [] globalanswer ;
    public int n,globaltarget;
    public int solution(int[] numbers, int target) {

        globalanswer = numbers;
        n = numbers.length;
        globaltarget = target;
        dfs(0,0);
        return answer;
    }
    
    public void dfs(int val,int depth)
    {
        if (depth == n)
        {
            if (val == globaltarget)
            {
                answer+=1;
 
            }              
            return;
        }
        
        dfs(val-globalanswer[depth],depth+1);
        dfs(val+globalanswer[depth],depth+1);
    }
}